// eslint-plugin-import/lib/core/resolve.js
'use strict';

var _slicedToArray = require('babel-runtime/helpers/sliced-to-array')['default'];

var _Map = require('babel-runtime/core-js/map')['default'];

var _getIterator = require('babel-runtime/core-js/get-iterator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports.relative = relative;
exports['default'] = resolve;

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _path = require('path');

var CASE_INSENSITIVE = _fs2['default'].existsSync((0, _path.join)(__dirname, 'reSOLVE.js'));

// http://stackoverflow.com/a/27382838
function fileExistsWithCaseSync(_x) {
  var _again = true;

  _function: while (_again) {
    var filepath = _x;
    _again = false;

    var dir = (0, _path.dirname)(filepath);
    if (dir === '/' || dir === '.' || /^[A-Z]:\\$/i.test(dir)) return true;
    var filenames = _fs2['default'].readdirSync(dir);
    if (filenames.indexOf((0, _path.basename)(filepath)) === -1) {
      return false;
    }
    _x = dir;
    _again = true;
    dir = filenames = undefined;
    continue _function;
  }
}

function fileExists(filepath) {
  if (CASE_INSENSITIVE) {
    // short-circuit if path doesn't exist, ignoring case
    return !(!_fs2['default'].existsSync(filepath) || !fileExistsWithCaseSync(filepath));
  } else {
    return _fs2['default'].existsSync(filepath);
  }
}

function relative(modulePath, sourceFile, settings) {

  function withResolver(resolver, config) {
    // resolve just returns the core module id, which won't appear to exist
    try {
      var filePath = resolver.resolveImport(modulePath, sourceFile, config);
      if (filePath === null) return null;

      if (filePath === undefined || !fileExists(filePath)) return undefined;

      return filePath;
    } catch (err) {
      return undefined;
    }
  }

  var configResolvers = settings['import/resolver'] || { 'node': settings['import/resolve'] }; // backward compatibility

  var resolvers = resolverReducer(configResolvers, new _Map());

  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = _getIterator(resolvers.entries()), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var _step$value = _slicedToArray(_step.value, 2);

      var _name = _step$value[0];
      var config = _step$value[1];

      var resolver = require('eslint-import-resolver-' + _name);

      var fullPath = withResolver(resolver, config);
      if (fullPath !== undefined) return fullPath;
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator['return']) {
        _iterator['return']();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }
}

function resolverReducer(resolvers, map) {
  if (resolvers instanceof Array) {
    resolvers.forEach(function (r) {
      return resolverReducer(r, map);
    });
    return map;
  }

  if (typeof resolvers === 'string') {
    map.set(resolvers, null);
    return map;
  }

  if (typeof resolvers === 'object') {
    for (var key in resolvers) {
      map.set(key, resolvers[key]);
    }
    return map;
  }

  throw new Error('invalid resolver config');
}

/**
 * Givent
 * @param  {string} p - module path
 * @param  {object} context - ESLint context
 * @return {string} - the full module filesystem path;
 *                    null if package is core;
 *                    undefined if not found
 */

function resolve(p, context) {
  return relative(p, context.getFilename(), context.settings);
}

resolve.relative = relative;
